Userland/Tests/Kernel/munmap-multi-region-unmapping.cpp

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Types.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+int main()
+{
+    const size_t size = PAGE_SIZE;
+    auto fd = open("/home/anon/Documents/emoji.txt", O_RDONLY);
+    if (fd < 0) {
+        perror("open");
+        return 1;
+    }
+    {
+        auto* map1 = mmap(nullptr, size, PROT_READ, MAP_FILE | MAP_SHARED, fd, 0);
+        if (map1 == MAP_FAILED) {
+            perror("mmap11");
+            return 1;
+        }
+        auto* map2 = mmap((void*)((FlatPtr)map1 + size), size, PROT_READ, MAP_FILE | MAP_SHARED, fd, 0);
+        if (map2 == MAP_FAILED) {
+            perror("mmap21");
+            return 1;
+        }
+
+        int res = munmap(map1, size * 2);
+        if (res < 0) {
+            perror("unmap1");
+            return 1;
+        }
+    }
+    {
+        auto* map1 = mmap(nullptr, size, PROT_READ, MAP_FILE | MAP_SHARED, fd, 0);
+        if (map1 == MAP_FAILED) {
+            perror("mmap12");
+            return 1;
+        }
+        auto* map2 = mmap((void*)((FlatPtr)map1 + size), size, PROT_READ, MAP_FILE | MAP_SHARED, fd, 0);
+        if (map2 == MAP_FAILED) {
+            perror("mmap22");
+            return 1;
+        }
+
+        int res = munmap((void*)((FlatPtr)map1 + size / 2), size);
```

Comment:
I don't think this does what you think this does. [posix(ish) says](https://linux.die.net/man/3/munmap):

> The munmap() function shall remove any mappings for those entire pages containing any part of the address space of the process starting at addr and continuing for len bytes.

So, this should remove all mappings that contain any part of the [map1 + half-page, map1 + page + half-page] interval. Therefore the very first thing that munmap should do is rounding the start and end to [map1, map1 + 2 * page], so this ends up testing the exact same thing as the previous testcase.

It's good that you test this scenario: After all, this is the tricky case that would have triggered the previous bugs.

One approach would be to just mmap 2 pages each; it doesn't need a filedescriptor, so I wouldn't use one here. For the first mmap, the invocation should probably look like this: `mmap(nullptr, 2 * PAGE_SIZE, PROT_READ, MAP_ANONYMOUS | MAP_PRIVATE, 0, 0)`. Make sure to write to the pages to force the Kernel to actually allocate the memory, and read from the valid pages after deallocating. Only this way we can verify that the memory that should still be valid (the first and last page) is actually still valid.

Something like this:
```
map1 = mmap(nullptr, 2 * PAGE_SIZE, PROT_READ, MAP_ANONYMOUS | MAP_PRIVATE, 0, 0)
map2 = mmap(map2 + 2 * PAGE_SIZE, 2 * PAGE_SIZE, PROT_READ, MAP_ANONYMOUS | MAP_PRIVATE | MAP_FIXED, 0, 0)
((int*)map1)[0] = 0x12345678;
((int*)map1)[PAGE_SIZE / sizeof(int)] = 0xabcdef;
((int*)map2)[0] = 0x87654321;
((int*)map2)[PAGE_SIZE / sizeof(int)] = 0xfdebca;
munmap(map1 + PAGE_SIZE, PAGE_SIZE * 2)
VERIFY(((int*)map1)[0] == 0x12345678);
VERIFY(((int*)map2)[PAGE_SIZE / sizeof(int)] == 0xfdebca);
```

(Bonus points if you figure out how to check that the munmap'd space has actually been unmapped. I don't know how to do that, so it's fine to leave it unchecked.)

Reply:
So we should round to pages when checking for intersections or when we lookup the regions intersecting? (if yes, wich ifnot both)

Reply:
To check if they were unmapped we could just try to remap it, can't we?

Reply:
> we could just try to remap it

Good idea!

Reply:
> So we should round to pages when checking for intersections or when we lookup the regions intersecting? (if yes, wich ifnot both)

So well actually we expand the range_to_unmap before we even start looking for regions

Reply:
The Tets does now checks all those things, and it worked without a hitch

Reply:
@rounding: The Kernel already does, it's just your test program that forgot about it. :P

In the Kernel, this job is done by the `auto range_or_error = expand_range_to_page_boundaries((FlatPtr)addr, size);` line.